web-access-control/protected-operation/read-access-bob.feature

```diff
@@ -1,119 +1,210 @@
 Feature: Only Bob can read (and only that) a resource when granted read access
   # Grant a specific agent (setAgentAccess):
-  # - full access to the parent container (to ensure the tests are specific to the resource)
-  # - restricted access to the test resources
+  # - restricted access or no access to the parent container
+  # - restricted access to the test resources, or inherited access for fictive resources
   Background: Create test resources with correct access modes
-    * def authHeaders = (method, url, public) => !public ? clients.bob.getAuthHeaders(method, url) : {}
-    * def createResources =
+    * def authHeaders =
+    """
+      function (method, url, agent) {
+        const agentLowerCase = agent.toLowerCase()
+        return agentLowerCase != 'public' ? clients[agentLowerCase].getAuthHeaders(method, url) : {}
+      }
+    """
+    * def resourcePermissions =
     """
       function (modes) {
+        if (modes && modes != 'inherited' && modes != 'no') {
+          return Object.entries({ R: 'read', W: 'write', A: 'append', C: 'control' })
+            .filter(([mode, permission]) => modes.includes(mode))
+            .map(([mode, permission]) => permission)
+        }
+        return undefined
+      }
+    """
+    * def resourceEntry =
+    """
+      function (container, type) {
+        switch (type) {
+          case 'plain':
+            return container.createResource('.txt', 'Hello', 'text/plain')
+          case 'fictive':
+            return container.reserveResource('.txt')
+          case 'rdf':
+            return container.createResource('.ttl', karate.readAsString('../fixtures/example.ttl'), 'text/turtle')
+          case 'container':
+            return container.createContainer()
+          default:
+            return undefined
+        }
+      }
+    """
+    * def createResource =
+    """
+      function (containerModes, resourceModes, resourceType, agent) {
+
+        const agentWebId = webIds.bob
+        const testContainerPermissions = resourcePermissions(containerModes == 'all' ? 'RWAC' : containerModes)
+        const testResourcePermissions = resourcePermissions(resourceModes)
+
+        const testContainerInheritablePermissions = resourceModes == 'inherited'
+          ? testContainerPermissions
+          : undefined
+
         const testContainer = rootTestContainer.createContainer()
+        const testResource = resourceEntry(testContainer, resourceType)
+
         testContainer.accessDataset = testContainer.accessDatasetBuilder
-          .setAgentAccess(testContainer.url, webIds.bob, ['read', 'write', 'append', 'control']).build()
```

Comment:
How do we ensure we are testing exactly what we expect? The original test deliberately granted Bob full access to the container (with nothing inheritable), then gave a specific mode to the child resource and tested whether access control did what we expected on that resource. So Bob **cannot** read even if given all other access modes, and Bob **can** read when given read but nothing else. I'm confirming that no accidental access is granted and also confirming that the tests are valid. I saw some previous tests where all read tests passed because the Alice user (owner) had been used instead of Bob so the tests were all invalid. Your test works a little differently. You grant no access to the container and read on the resource then check Bob can read. I don't see the equivalent of my other test. Did I miss it? What are your thoughts on that?

I also think there is a problem here when the container permissions are "no". The value of `testContainerPermissions` ends up undefined but you still use that to call `setAgentAccess` (similar with `setInheritableAgentAccess`). It results in an incomplete ACR. Perhaps I should change the test harness to allow undefined to be passed to the method and then don't create any policy for that agent rather than create a policy with no access modes. However, whether or not that is needed may be impacted by what we agree for the first part of this comment.

Reply:
I think this is related to the comment above, with the leftover `all` permissions, that I ended up not using. There are indeed no tests with all permissions on container and limited permissions on resource, as there were originally. Maybe that is problematic, I just did not view it that way, being relatively new to these things still. My apologies.

In the original tests, when checking if Bob can read a resource (with permissions on the resource), all permissions were granted on the container (with no inherited ones) and only read permissions were granted on the resource. In the new ones, no permissions are granted on the container and only the permission being tested is granted on the resource.

The AWC but no R tests I moved to the `write-access-{agent,bob,public}.feature` because those were concerned with write permissions. But the similar changes to container vs resource permissions were done there, as well.

Would it make sense to revert the tests back to assigning all permissions on container (not inherited) when testing on resources alone?

That last part about undefined being used seems like a bug indeed, in my resource setup script, and I will fix it. When no permissions are assigned, the function should not be called. I forgot to add the checks. Thank you!

Reply:
It seems I cannot reply to the topmost message, so I will write it here instead. For JavaScript, I will indeed need to change the comparison operators, I keep forgetting that. Thank you.

The tests are maybe laid out on a confusing way, but I have tried to add tests for:
* R in `read-access-{agent,bob,public}.feature`
* WAC in `write-access-{agent,bob,public}.feature`, sometimes splitting W, A and C, as they were always handled separately in the CSS tests (to check that they are not linked, I assume, by implementation?)
* Directly on resource, with none on container, so that an implementation does not (mistakenly) have resources inherit container permissions when not supposed to (probably really unlikely, but can never know)
* Inherited from container, with none on resource, and always like this for fictive ones since they cannot have permissions on the resource

I can try to explain the reasoning better or in more detail when I know what to explain! :slightly_smiling_face:

Reply:
I've been working through your read tests and I have a better understanding of what you have done now, and I think it is great. Your logic on setting the resource or inherited container permissions makes sense to me. I have extracted the javascript functions at the start of each read tests into a common feature which is shared between the other features. I also reworked the access dataset setup to avoid creating invalid policies when the permissions are undefined. This is all looking great so tomorrow I'll try and do the same to the write tests and then see if I can push those changes to this PR.